### PR TITLE
draft of get, read and refactor to readOrRegisterResource

### PR DESCRIPTION
**Pull request by lbialy**

### CODE REVIEW

**Review by pawelprazak — APPROVED** [review]

File: core/src/main/scala/besom/internal/ResourceOptions.scala [hunk 1]
```diff
@@ -44,13 +46,25 @@ sealed trait ResourceOptions:
...
+    case sr: StackReferenceResourceOptions => sr.importId
+    case _                                 => None
 
 final case class CustomResourceOptions private[internal] (
   common: CommonResourceOptions,
   provider: Option[ProviderResource],
   deleteBeforeReplace: Boolean,
   additionalSecretOutputs: List[String],
-  importId: Option[NonEmptyString] // TODO should this be Id?
+  importId: Option[ResourceId] // TODO should this be Id?
```
Comment (pawelprazak) [inline core/src/main/scala/besom/internal/ResourceOptions.scala:67]: nit: the todo note appears to be fixed now :)

File: core/src/main/scala/besom/internal/ResourceOptions.scala [hunk 2]
```diff
@@ -81,13 +95,13 @@ trait CustomResourceOptionsFactory:
     customTimeouts: String | NotProvided = NotProvided, // CustomTimeouts // TODO
     // resourceTransformations: List[ResourceTransformation], // TODO
     // aliases: List[Output[Alias]], // TODO
-    urn: String | NotProvided = NotProvided, // TODO better type
+    urn: URN | NotProvided = NotProvided, // TODO better type
```
Comment (pawelprazak) [inline core/src/main/scala/besom/internal/ResourceOptions.scala:98]: nit: the todo appear to be fixed now

File: core/src/main/scala/besom/types.scala [hunk 3]
```diff
@@ -157,21 +163,61 @@ object types:
...
 
-    // This is implemented according to https://www.pulumi.com/docs/concepts/resources/names/#urns
-    private[types] val UrnRegex =
-      """urn:pulumi:(?<stack>[^:]+)::(?<project>[^:]+)::(?<parentType>.+?)((\$(?<resourceType>.+))?)::(?<resourceName>.+)""".r
+    /** This is implemented according to https://www.pulumi.com/docs/concepts/resources/names/#urns and
+      * https://pulumi-developer-docs.readthedocs.io/en/latest/providers/implementers-guide.html?highlight=URN#urns
+      *
+      * If you want to understand this regex better head to: https://regex101.com/r/o2QWJ3/1
+      *
+      * Now hear me out, this regex is a bit complicated, but it's not that bad. Let's break it down. We have to adhere to this grammar:
```
Comment (pawelprazak) [inline core/src/main/scala/besom/types.scala:173]: I like your humor, "not that bad regex" is dark, but still funny ;)

Don't worry, one day we'll write a proper parser :)

File: integration-tests/integration.scala [hunk 4]
```diff
@@ -121,44 +123,117 @@ object pulumi {
...
-      val stackName     = testToStack(test.name)
-      val tmpPulumiHome = os.temp.dir()
+    ): munit.TestOptions => FixtureContext =
+      val pulumiContext  = init(FixtureOpts(pulumiHomeDir, pulumiEnv))
+      val programContext = setup(pulumiContext, FixtureArgs(testDir, projectFiles))
+      (test: munit.TestOptions) => FixtureContext(pulumiContext, programContext(test))
+
+    def init(opts: FixtureOpts): PulumiContext = {
+      val allEnv: Map[String, String] =
+        // local environment, to run locally offline, make sure you set PULUMI_BACKEND_URL and PULUMI_API
```
Comment (pawelprazak) [inline integration-tests/integration.scala:187]:
```suggestion
```

File: integration-tests/integration.scala [hunk 5]
```diff
@@ -121,44 +123,117 @@ object pulumi {
...
+        opts.pulumiEnv ++ Map(
+          //          "PULUMI_BACKEND_URL" -> s"file://${opts.pulumiHomeDir.toString}",
+          //          "PULUMI_API" -> s"file://${opts.pulumiHomeDir.toString}",
```
Comment (pawelprazak) [inline integration-tests/integration.scala:190]:
```suggestion
```

File: integration-tests/CodegenTests.test.scala [hunk 6]
```diff
@@ -31,24 +31,24 @@ class CodegenTests extends munit.FunSuite {
 
   // FIXME: broken - codegen error
   val ignoreList = List(
-    "simple-resource-schema", // resource decoder issue
+    // "simple-resource-schema", // fixed
```
Comment (pawelprazak) [inline integration-tests/CodegenTests.test.scala:34]:
```suggestion
```

File: integration-tests/CodegenTests.test.scala [hunk 7]
```diff
@@ -31,24 +31,24 @@ class CodegenTests extends munit.FunSuite {
...
     "simple-enum-schema", // simple enum is not supported
     "simple-yaml-schema", // YAML is not supported
     "external-enum", // depends on google-native, TODO: check if this is still broken
-    "external-resource-schema", // resource decoder issue
+    // "external-resource-schema", // fixed
```
Comment (pawelprazak) [inline integration-tests/CodegenTests.test.scala:38]:
```suggestion
```

File: integration-tests/CodegenTests.test.scala [hunk 8]
```diff
@@ -31,24 +31,24 @@ class CodegenTests extends munit.FunSuite {
...
     "enum-reference", // depends on google-native, TODO: check if this is still broken
     "different-enum", // simple enum is not supported
     "hyphen-url", // depends on azure-native,
     "naming-collisions", // codec not found
     "mini-azurenative", // simple enum is not supported
-    "replace-on-change", // resource decoder issue
-    "resource-property-overlap", // resource decoder issue
+    // "replace-on-change", // fixed
+    // "resource-property-overlap", // fixed
```
Comment (pawelprazak) [inline integration-tests/CodegenTests.test.scala:45]:
```suggestion
```

File: core/src/main/scala/besom/internal/ResourceOptions.scala (hunk 2 above)
Comment (pawelprazak) [reply core/src/main/scala/besom/internal/ResourceOptions.scala:98]:
```suggestion
    urn: URN | NotProvided = NotProvided,
```

File: core/src/main/scala/besom/internal/ResourceOptions.scala (hunk 1 above)
Comment (pawelprazak) [reply core/src/main/scala/besom/internal/ResourceOptions.scala:67]:
```suggestion
  importId: Option[ResourceId]
```

Comment (lbialy) [reply core/src/main/scala/besom/internal/ResourceOptions.scala:67]: it isn't really, resource options require a rewrite

**Pull request merged**